File: backend/app/models/category.py
"""
Category Model

Represents a transaction category (income or expense)
"""
import enum
import uuid
from datetime import datetime

from sqlalchemy import Column, DateTime, ForeignKey, String
from sqlalchemy import Enum as SQLEnum
from sqlalchemy.orm import relationship

from app.database import Base


class CategoryType(str, enum.Enum):
    """Category types"""
    INCOME = "INCOME"  # Revenu
    EXPENSE = "EXPENSE"  # Dépense


class Category(Base):
    """Category model"""
    __tablename__ = "categories"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    household_id = Column(String(36), ForeignKey("households.id", ondelete="CASCADE"), nullable=False)
    name = Column(String(100), nullable=False)
    type = Column(SQLEnum(CategoryType), nullable=False)
    icon = Column(String(50), nullable=True)  # Emoji or icon name
    color = Column(String(7), nullable=True)  # Hex color (e.g., #FF5733)
    parent_id = Column(String(36), ForeignKey("categories.id", ondelete="CASCADE"), nullable=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    household = relationship("Household", back_populates="categories")
    parent = relationship("Category", remote_side=[id], backref="children")
    transactions = relationship("Transaction", back_populates="category")

    def __repr__(self):
        return f"<Category {self.name} ({self.type})>"

File: frontend/src/types/category.ts
/**
 * Category Types
 */

export enum CategoryType {
  INCOME = "INCOME",
  EXPENSE = "EXPENSE",
}

export interface Category {
  id: string;
  household_id: string;
  name: string;
  type: CategoryType;
  icon: string | null;
  color: string | null;
  parent_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface CategoryCreate {
  name: string;
  type: CategoryType;
  icon?: string;
  color?: string;
  parent_id?: string;
}

export interface CategoryUpdate {
  name?: string;
  icon?: string;
  color?: string;
  parent_id?: string;
}

export interface CategoryTree extends Category {
  children: CategoryTree[];
}

export const CATEGORY_TYPE_LABELS: Record<CategoryType, string> = {
  [CategoryType.INCOME]: "Revenu",
  [CategoryType.EXPENSE]: "Dépense",
};

export const DEFAULT_CATEGORY_COLORS = [
  "#27AE60", // Green
  "#E74C3C", // Red
  "#3498DB", // Blue
  "#F39C12", // Orange
  "#9B59B6", // Purple
  "#1ABC9C", // Turquoise
  "#E67E22", // Dark Orange
  "#34495E", // Dark Gray
  "#16A085", // Dark Turquoise
  "#C0392B", // Dark Red
];

export const DEFAULT_CATEGORY_ICONS = [
  "🏠", "🍔", "🚗", "🎮", "💊", "🎓", "✈️", "🛒",
  "💰", "💳", "📱", "⚡", "🎬", "🏋️", "🎨", "📚",
];
